### DISCUSSION TITLE
How to clone a dataset and exclude the pixel data

### DISCUSSION THREAD

**Discussion by Tony-Jay** [fo-dicom/fo-dicom#1298] [Q&A]
Hi,

Is there an elegant way to clone just the header information from a Dataset ?

I have datasets which are multiimage 100's of MB in size (breast Tomo etc)

I want to operate on one  dataset (modifying some of the tags to deidentify it using our in house standards) in order to do that I need to keep a copy of the header data as deidentifying some items require other items to keep dates/ ages coherent

To make it worse I am actually a library  not an application which receiver a DicomFile

Inelegant options
Option A
1. clone the dataset (taking up 100s of MB of memory per dataset - and this could be multiple threads with multiple datasets)
2. delete the pixel data of the cloned dataset

Option B
1. write the DicomFile to Disk
2. delete the pixel data of the datset in memory
3. read back in the whole dataset with pixel data

Option C
What I would prefer is an extension method "CloneOnlyHeaderData" , something like

         1. copy meta data to new dicom file of right tr
         2. get enumerable Tags from old dataset  (iterate on original.DataSet) 
         3. remove pixeldatatag from enumerable
         4. add enumerable to dataset of new file , created in correct transfer syntax
         
            
Any comments on my proposed  approach ?  (C)

regards

Tony

**Comment by gofal** [comment]
Option C would be the best method, so if you create a PR this would be very welcome!
This method could solve a few issues: for example if you just remove the pixeldata, then you get an invalid file. This extension method should set the number of frames to 0, for example. it should also get a new SopInstanceUID and save a reference to the original SopInstanceUID in "Referenced SOP Sequence". and so on. This way you would not do some hack, but you would really create a new valid dicomfile that contains the required information in much smaller size.

But until that is done, Option A should be fine.
fo-dicom has to clone datasets very often internally when doing all the network communication. therefore cloning has been optimized. The DicomItems are immutable objects and a DicomDataset is a list pointing to such DicomItems. When you clone a DicomDataset, then you get a new List, but this list is pointing to exactly the same DicomItem-instances in memory. The DicomItem-instances are not duplicated. In other words, the Clone method does not do a deep cloning but only a shallow cloning.
This is save, because of this immutable design. Maybe you noticed, that you cannot change the value of an item directly (e.g. when you want to update the PatientName for anonymization), but when you call "DicomDataset.AddOrUpdate", then a new DicomItem with the new value is created, and the reference to the old DicomItem is updated to reference the new DicomItem. 

When you clone the DicomDataset, you should NOT notice a huge increase of memory consumption. if you do so, then open an issue and post some sample code - this has to be fixed.

**Comment by Tony-Jay** [comment]
Thanks a lot, the explanation about the shallow copy is is really useful.

When I had a look at memory usage I was also changing transfer syntax (from jpeg  ) and got a  huge file so that is why I (incorrectly) assumed it was a deep copy - obviously changing transfer syntax is not the same as just a clone. I will write some tests with clone and large datasets.

With your information, I will see if the normal Clone is good enough for my use case (rather complicated deidentification strategy allowing original data to be reidentified)
